File: backend/matching-engine/src/main/java/com/commonsengine/matching/api/MatchingController.java
package com.commonsengine.matching.api;

import com.commonsengine.matching.engine.MatchingEngine;
import com.commonsengine.matching.infrastructure.persistence.WorkerLocationEntity;
import com.commonsengine.matching.service.WorkerLocationService;
import com.commonsengine.platform.domain.Model.MatchResult;
import com.commonsengine.platform.domain.Model.ServiceRequest;
import com.commonsengine.platform.domain.Model.Worker;
import com.commonsengine.platform.domain.Model.WorkerId;
import com.commonsengine.platform.domain.Model.ConsumerId;
import com.commonsengine.platform.domain.Model.RequestId;
import com.commonsengine.platform.domain.ServiceType;
import com.commonsengine.platform.geo.GeoPoint;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * 匹配 REST API（#62 #63）
 *
 * 端点：
 * - GET    /api/v1/matching/strategies           列出可用策略
 * - GET    /api/v1/matching/strategy              查询当前策略
 * - POST   /api/v1/matching/strategy              切换策略
 * - POST   /api/v1/matching/match                 手动匹配
 * - POST   /api/v1/matching/auto-match            自动匹配（基于附近劳动者）
 * - POST   /api/v1/matching/locations/{workerId}  更新劳动者位置
 * - GET    /api/v1/matching/nearby                查询附近劳动者
 */
@RestController
@RequestMapping("/api/v1/matching")
public class MatchingController {

    private final MatchingEngine engine;
    private final WorkerLocationService locationService;

    public MatchingController(MatchingEngine engine, WorkerLocationService locationService) {
        this.engine = engine;
        this.locationService = locationService;
    }

    @GetMapping("/strategies")
    public List<String> strategies() {
        return List.copyOf(engine.availableStrategies());
    }

    @GetMapping("/strategy")
    public StrategyRequest currentStrategy() {
        return new StrategyRequest(engine.currentStrategy());
    }

    @PostMapping("/strategy")
    public StrategyRequest switchStrategy(@RequestBody StrategyRequest request) {
        engine.useStrategy(request.name());
        return new StrategyRequest(engine.currentStrategy());
    }

    /**
     * 手动匹配——请求中直接提供候选劳动者列表。
     */
    @PostMapping("/match")
    public ResponseEntity<MatchResponse> match(@RequestBody MatchRequest request) {
        ServiceRequest serviceRequest = buildServiceRequest(request);
        List<Worker> candidates = buildWorkers(request.candidates());

        MatchResult result = engine.match(serviceRequest, candidates);
        if (result == null) {
            return ResponseEntity.ok(MatchResponse.noMatch(engine.currentStrategy()));
        }
        return ResponseEntity.ok(MatchResponse.from(result));
    }

    /**
     * 自动匹配——基于请求位置查询附近劳动者后匹配。
     */
    @PostMapping("/auto-match")
    public ResponseEntity<MatchResponse> autoMatch(@RequestBody AutoMatchRequest request) {
        List<WorkerLocationEntity> nearby = locationService.findNearbyWorkers(
                request.pickupLat(), request.pickupLng(), request.radiusMeters()
        );
        if (nearby.isEmpty()) {
            return ResponseEntity.ok(MatchResponse.noMatch(engine.currentStrategy()));
        }

        ServiceRequest serviceRequest = buildServiceRequestFromAuto(request);
        List<Worker> workers = new ArrayList<>();
        for (WorkerLocationEntity e : nearby) {
            workers.add(buildWorker(e.getWorkerId(), e.getLat(), e.getLng(), e.getActiveOrderCount()));
        }

        MatchResult result = engine.match(serviceRequest, workers);
        if (result == null) {
            return ResponseEntity.ok(MatchResponse.noMatch(engine.currentStrategy()));
        }
        return ResponseEntity.ok(MatchResponse.from(result));
    }

    @PostMapping("/locations/{workerId}")
    public LocationUpdate updateLocation(@PathVariable String workerId,
                                          @RequestBody LocationUpdate update) {
        locationService.updateLocation(workerId, update.lat(), update.lng(),
                update.activeOrderCount());
        return update;
    }

    @GetMapping("/nearby")
    public List<CandidateDto> nearby(@RequestParam double lat,
                                      @RequestParam double lng,
                                      @RequestParam(defaultValue = "5000") double radiusMeters) {
        return locationService.findNearbyWorkers(lat, lng, radiusMeters).stream()
                .map(e -> new CandidateDto(e.getWorkerId(), e.getLat(), e.getLng(),
                        e.getActiveOrderCount()))
                .toList();
    }

    // ── 构造 Java 领域对象 ────────────────────────────────
    // platform-core 域模型为 Java record（JDK 21），ID 为 record 包装类型。

    private ServiceRequest buildServiceRequest(MatchRequest req) {
        return new ServiceRequest(
                RequestId.random(),
                new ConsumerId(req.consumerId()),
                req.serviceType(),
                new GeoPoint(req.pickupLat(), req.pickupLng()),
                new GeoPoint(req.pickupLat(), req.pickupLng())
        );
    }

    private ServiceRequest buildServiceRequestFromAuto(AutoMatchRequest req) {
        return new ServiceRequest(
                RequestId.random(),
                new ConsumerId(req.consumerId()),
                req.serviceType(),
                new GeoPoint(req.pickupLat(), req.pickupLng()),
                new GeoPoint(req.pickupLat(), req.pickupLng())
        );
    }

    private List<Worker> buildWorkers(List<CandidateDto> candidates) {
        List<Worker> workers = new ArrayList<>();
        for (CandidateDto c : candidates) {
            workers.add(buildWorker(c.workerId(), c.lat(), c.lng(), c.activeOrderCount()));
        }
        return workers;
    }

    private Worker buildWorker(String workerId, double lat, double lng, int activeOrderCount) {
        return new Worker(
                new WorkerId(workerId),
                workerId,
                new GeoPoint(lat, lng),
                5.0,
                activeOrderCount,
                Set.of()
        );
    }

    // ── DTO records ────────────────────────────────────────

    public record StrategyRequest(@NotBlank String name) {
    }

    public record MatchRequest(
            @NotBlank String consumerId,
            @NotNull ServiceType serviceType,
            double pickupLat,
            double pickupLng,
            List<CandidateDto> candidates
    ) {
    }

    public record AutoMatchRequest(
            @NotBlank String consumerId,
            @NotNull ServiceType serviceType,
            double pickupLat,
            double pickupLng,
            double radiusMeters
    ) {
    }

    public record CandidateDto(
            @NotBlank String workerId,
            double lat,
            double lng,
            int activeOrderCount
    ) {
    }

    public record LocationUpdate(double lat, double lng, int activeOrderCount) {
    }

    public record MatchResponse(
            boolean matched,
            String workerId,
            Double score,
            String strategy,
            String reason
    ) {
        static MatchResponse from(MatchResult r) {
            // MatchResult 是 record(workerId, requestId, score)
            return new MatchResponse(
                    true,
                    r.workerId().toString(),
                    r.score(),
                    null,
                    "匹配成功"
            );
        }

        static MatchResponse noMatch(String strategy) {
            return new MatchResponse(false, null, null, strategy, "无匹配候选");
        }
    }
}

File: backend/matching-engine/src/main/kotlin/com/commonsengine/matching/api/MatchingController.kt
package com.commonsengine.matching.api

import com.commonsengine.matching.engine.MatchingEngine
import com.commonsengine.matching.service.WorkerLocationService
import com.commonsengine.platform.domain.ConsumerId
import com.commonsengine.platform.domain.RequestId
import com.commonsengine.platform.domain.ServiceRequest
import com.commonsengine.platform.domain.ServiceType
import com.commonsengine.platform.domain.Worker
import com.commonsengine.platform.domain.WorkerId
import com.commonsengine.platform.geo.GeoPoint
import com.commonsengine.platform.support.Enums
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

/**
 * 匹配引擎 REST API
 */
@RestController
@RequestMapping("/api/v1/matching")
open class MatchingController(
    private val engine: MatchingEngine,
    private val locationService: WorkerLocationService,
) {

    /** 健康检查 + 当前策略 */
    @GetMapping("/health")
    fun health(): Map<String, Any> = mapOf(
        "status" to "UP",
        "currentStrategy" to engine.currentStrategy(),
        "availableStrategies" to engine.availableStrategies(),
    )

    /** 切换匹配策略 */
    @PostMapping("/strategy")
    fun setStrategy(@Valid @RequestBody body: StrategyRequest): Map<String, String> {
        engine.setStrategy(body.strategy)
        return mapOf("status" to "ok", "currentStrategy" to engine.currentStrategy())
    }

    /**
     * 手动匹配——调用方提供候选列表（兼容旧接口）
     */
    @PostMapping("/match")
    fun match(@Valid @RequestBody body: MatchRequest): MatchResponse {
        val request = toServiceRequest(body)
        val candidates = body.candidates.map { it.toWorker() }
        return doMatch(request, candidates)
    }

    /**
     * 自动匹配——系统从数据库检索附近劳动者（#37 PostGIS 地理索引）
     *
     * 调用方只需提供位置和服务类型，系统自动查找 radiusMeters 范围内的劳动者。
     */
    @PostMapping("/match/auto")
    fun autoMatch(@Valid @RequestBody body: AutoMatchRequest): MatchResponse {
        val request = ServiceRequest(
            id = RequestId.random(),
            consumerId = ConsumerId(body.consumerId),
            type = Enums.parse<ServiceType>(body.serviceType),
            pickupLocation = GeoPoint(body.pickupLat, body.pickupLng),
        )

        val candidates = locationService.findNearbyWorkers(
            center = request.pickupLocation,
            radiusMeters = body.radiusMeters ?: 5000.0,
            maxActiveOrders = body.maxActiveOrders ?: 3,
        )

        return doMatch(request, candidates)
    }

    /**
     * 劳动者上报位置（心跳）
     */
    @PostMapping("/workers/{workerId}/location")
    fun updateLocation(
        @PathVariable workerId: String,
        @Valid @RequestBody body: LocationUpdate,
    ): Map<String, String> {
        locationService.upsertLocation(
            workerId = workerId,
            name = body.name,
            lat = body.lat,
            lng = body.lng,
            serviceTypes = body.serviceTypes?.joinToString(",") ?: "",
            rating = body.rating ?: 5.0,
            activeOrderCount = body.activeOrderCount ?: 0,
        )
        return mapOf("status" to "ok")
    }

    // ── 内部方法 ──────────────────────────────────────

    private fun toServiceRequest(body: MatchRequest) = ServiceRequest(
        id = RequestId.random(),
        consumerId = ConsumerId(body.consumerId),
        type = Enums.parse<ServiceType>(body.serviceType),
        pickupLocation = GeoPoint(body.pickupLat, body.pickupLng),
    )

    private fun doMatch(request: ServiceRequest, candidates: List<Worker>): MatchResponse {
        val result = engine.match(request, candidates)
            ?: return MatchResponse(
                matched = false,
                reason = "无合格候选劳动者（候选数: ${candidates.size}）",
                strategy = engine.currentStrategy(),
            )

        return MatchResponse(
            matched = true,
            workerId = result.worker.id.value,
            workerName = result.worker.name,
            distanceMeters = result.distanceMeters,
            strategy = result.strategy,
            reason = result.reason,
        )
    }
}

// ── DTO ──────────────────────────────────────────────────────────

data class StrategyRequest(@field:NotBlank val strategy: String)

data class MatchRequest(
    @field:NotBlank val consumerId: String,
    @field:NotBlank val serviceType: String,
    val pickupLat: Double,
    val pickupLng: Double,
    val candidates: List<CandidateDto>,
)

data class AutoMatchRequest(
    @field:NotBlank val consumerId: String,
    @field:NotBlank val serviceType: String,
    val pickupLat: Double,
    val pickupLng: Double,
    val radiusMeters: Double? = null,
    val maxActiveOrders: Int? = null,
)

data class CandidateDto(
    val id: String,
    val name: String,
    val lat: Double,
    val lng: Double,
    val rating: Double = 5.0,
    val activeOrderCount: Int = 0,
) {
    fun toWorker() = Worker(
        id = WorkerId(id),
        name = name,
        currentLocation = GeoPoint(lat, lng),
        rating = rating,
        activeOrderCount = activeOrderCount,
    )
}

data class LocationUpdate(
    @field:NotBlank val name: String,
    val lat: Double,
    val lng: Double,
    val serviceTypes: List<String>? = null,
    val rating: Double? = null,
    val activeOrderCount: Int? = null,
)

data class MatchResponse(
    val matched: Boolean,
    val workerId: String? = null,
    val workerName: String? = null,
    val distanceMeters: Double? = null,
    val strategy: String,
    val reason: String,
)
